### PR TITLE
Update to Angular 15 and Javalin v5

### PR DESCRIPTION
**Pull request by jpwalbran**
The main purpose of this PR is to go through the update process to Angular 15, and then adjust anything that comes up as necessary. In the process of doing that, there were some other big things that came up and wound up in this PR.

This PR now includes:

- Updating Angular from 14 to 15
- Removing the Angular flex-layout
- Updating Angular Material to the newer MDC specifications (and updating Cypress e2e tests as needed to address these changes)
- Updating from Javalin v4 to v5 (and updating the tests for UserController)

The change to Javalin v5 was substantial, and required quite an overhaul of the tests. The UserControllerSpec in the iteration template matches more with what the students did in Lab 2 and includes use of Mockito as a replacement for the ContextUtil (which went away as a result of Context becoming an interface in v5 of Javalin). The challenges of matching actual usage of Javalin in the Mockito tests were significant. We probably do want to consider whether or not we like this type of test or if we would like to move to more functional testing like @NicMcPhee did with Unirest in this other branch of another repo: https://github.com/UMM-CSci-3601/3601-lab2_client-server/pull/224

This PR leaves the client looking a bit scraggly, but there is a PR in motion for fixing that somewhat: #1287

### CODE REVIEW

**Comment by kklamberty** [comment]
Ideally, we would also make the Angular stuff look a bit nicer before we start splitting out the elements for the labs. This is especially true for lab 3. We've removed flex-layout, but we haven't really dealt completely with the fallout of that choice. For example, the user list view is not looking particularly pretty yet. 

One reason to go ahead and merge in this branch is that is is focused on the updating of versions. We could have other branches that focus on separate details like making some changes that either don't rely on flex-box or could replace flex-box. There are some specifics about that in issue #1246 that are really not addressed here.

File: client/cypress/support/app.po.ts
Comment (NicMcPhee) [inline client/cypress/support/app.po.ts:18]: Why do you put spaces around these? That seems odd, and not something I would expect people (students) to do in the future? Is there a reason for it?

File: client/package.json
Comment (NicMcPhee) [inline client/package.json:33]: We need to update the `@angular-eslint` stuff to v15 as well, don't we?

File: client/package.json
Comment (NicMcPhee) [inline client/package.json:21]: Nice job getting rid of `flex-layout`.

File: client/src/app/home/home.component.html
Comment (NicMcPhee) [inline client/src/app/home/home.component.html:3]: Was this (and all the other additions of `appearance="outlined"`) needed as part of the update, or was it seen as an improvement to the interface? (Without running the previous version next to this one, it's hard to see what [if anything] this does.)

File: client/src/app/users/user-list.component.scss
Comment (NicMcPhee) [inline client/src/app/users/user-list.component.scss:15]: Nice change. Do we use `px` anywhere else?

File: client/src/test.ts
Comment (NicMcPhee) [inline client/src/test.ts:10]: Why don't we need this anymore? Is the default (somewhere) now to pick up all `.spec` files?

File: server/src/main/java/umm3601/user/User.java
Comment (NicMcPhee) [inline server/src/main/java/umm3601/user/User.java:47]: Good catch on the need to override the `hashCode` – sorry I missed that.

File: server/src/main/java/umm3601/user/UserController.java
Comment (NicMcPhee) [inline server/src/main/java/umm3601/user/UserController.java:115]: Nice job adding the validation to these requests.

**Review by NicMcPhee — CHANGES_REQUESTED** [review]
I really don't think you should merge this (as it claims to be about updating to Angular v15) until you also update the Angular `eslint` bits to v15.

My other comments are more cosmetic, or just straight up questions.

I agree 100% with @kklamberty's comment that we should _not_ let things like tidying up the appearance block this PR. Things like that are important, but should be in different PR(s) so we can get this merged in.

**Comment by NicMcPhee** [comment]
You should also update the PR description so that it more accurately reflects what all is actually happening in this PR, as it's grown well beyond just upgrading Angular to v15.

File: client/src/app/users/user-list.component.scss
Comment (kklamberty) [reply client/src/app/users/user-list.component.scss:15]: I think we use px *everywhere* else. I tried this kind of on a whim...

File: client/package.json
Comment (kklamberty) [reply client/package.json:21]: I'm not convinced this is actually ideal. The appearance of the cards and the lists is quite messed up by it. I was *just* about to make a commit that goes back to flex-layout so that we at least have cards that look decent. The list is still very broken either way. I'm not convinced that I have enough time before Lab 3 on *Tuesday* to do anything very good here. @NicMcPhee - Thoughts?

File: server/src/main/java/umm3601/user/User.java
Comment (kklamberty) [reply server/src/main/java/umm3601/user/User.java:47]: No worries!

File: client/package.json
Comment (NicMcPhee) [reply client/package.json]: I upgraded this in [50ad4ad](https://github.com/UMM-CSci-3601/3601-iteration-template/pull/1280/commits/50ad4adae0c981cfe7fa9a60d0f5ccae28049c5f), so that's good now.

File: client/cypress/support/app.po.ts
Comment (kklamberty) [reply client/cypress/support/app.po.ts:18]: I wonder if this happened with some of the automated tooling for moving to Angular 15? Or, possibly @jpwalbran made a choice here?

**Review by kklamberty — APPROVED** [review]

**Review by kklamberty — DISMISSED** [review]

File: client/src/app/home/home.component.html
Comment (kklamberty) [reply client/src/app/home/home.component.html:3]: It did look a little bit like you could not discern the location of the edges of the cards when we removed (and had not replaced) flex-layout, so I suspect this explains the reasoning behind this change.

File: client/src/app/users/user-list.component.scss
Comment (floogulinc) [reply client/src/app/users/user-list.component.scss:15]: The Angular Material stuff generally uses px values for styles are least as of the last time I checked.

File: .github/workflows/e2e.yaml
Comment (floogulinc) [inline .github/workflows/e2e.yaml:38]: If you're switching to Node 18 here then you should update it in the other workflows and the Dockerfile.

**Comment by floogulinc** [comment]
In the future when using `ng update` please use the `--create-commits` flag so it creates separate commits for each migration. This makes it much easier to see what was changes and why since each one has a comment explaining the change.

**Comment by NicMcPhee** [comment]
I did that, for example, on 50ad4adae0c981cfe7fa9a60d0f5ccae28049c5f. To be fair, those commit messages aren't amazing and don't read super well on GitHub, but they're consistent and do list all the updates, which is nice.

**Comment by kklamberty** [comment]
> In the future when using `ng update` please use the `--create-commits` flag so it creates separate commits for each migration. This makes it much easier to see what was changes and why since each one has a comment explaining the change.

Good call! I guess we just followed the instructions on the page about the migration and didn't also add that, but I will try to remember that for the future. Thanks for for the note!

File: .github/workflows/e2e.yaml
Comment (kklamberty) [reply .github/workflows/e2e.yaml:38]: Good catch! I definitely would not have remembered the Dockerfile.

**Comment by kklamberty** [comment]
Hey, All! I think I've addressed the changes requested. @NicMcPhee - can you please re-review? We need an approval from you to do this merge.

**Review by NicMcPhee — APPROVED** [review]
Thanks for all the good work - this was a bit of a bear.

**Pull request merged**
